File: 25. Reverse Nodes in k-Group/main.cpp
#include <algorithm>
#include <vector>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode()
        : val(0)
        , next(nullptr) {}
    ListNode(int x)
        : val(x)
        , next(nullptr) {}
    ListNode(int x, ListNode *next)
        : val(x)
        , next(next) {}
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    void reverse(vector<int> &values, int left, int right) {
        while (left < right) {
            swap(values[left++], values[right--]);
        }
    }

public:
    ListNode *reverseKGroup(ListNode *head, int k) {
        vector<int> values;
        ListNode *node = head;

        while (node) {
            values.push_back(node->val);
            node = node->next;
        }

        int size = values.size();

        for (int i = 0; i + k <= size; i += k) {
            reverse(values, i, i + k - 1);
        }

        node = head;
        for (int i = 0; i < size; i++) {
            node->val = values[i];
            node = node->next;
        }

        return head;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    ListNode **reverseList(ListNode **head, int k) {
        int count = 0;
        ListNode *current = *head;
        ListNode *prev = NULL;
        ListNode *next = NULL;

        while (current) {
            next = current->next;
            current->next = prev;
            prev = current;
            count++;
            if (count >= k) {
                break;
            }
            current = next;
        }

        ListNode *tail = *head;
        tail->next = next;
        *head = current;

        return &tail->next;
    }

public:
    ListNode *reverseKGroup(ListNode *head, int k) {
        int count = 0;
        ListNode *node = head;

        while (node) {
            count++;
            node = node->next;
        }

        ListNode **current = &head;

        while (*current) {
            if (count < k) {
                break;
            }

            current = reverseList(current, k);
            count -= k;
        }

        return head;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *reverseKGroup(ListNode *head, int k) {
        if (!head || k <= 1) {
            return head;
        }

        ListNode *node = head;

        for (int i = 0; i < k; i++) {
            if (!node) {
                return head;
            }
            node = node->next;
        }

        ListNode *current = head;
        ListNode *prev = NULL;
        ListNode *next = NULL;
        int count = 0;

        while (current && count < k) {
            next = current->next;
            current->next = prev;
            prev = current;
            current = next;
            count++;
        }

        if (next) {
            head->next = reverseKGroup(next, k);
        }

        return prev;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    ListNode *getKth(ListNode *node, int k) {
        while (node && k > 0) {
            node = node->next;
            k--;
        }

        return node;
    }

public:
    ListNode *reverseKGroup(ListNode *head, int k) {
        ListNode dummy = {0, head};
        ListNode *groupPrev = &dummy;

        while (true) {
            ListNode *kth = getKth(groupPrev, k);
            if (!kth) {
                break;
            }

            ListNode *groupNext = kth->next;
            ListNode *prev = kth->next;
            ListNode *current = groupPrev->next;

            while (current != groupNext) {
                ListNode *next = current->next;
                current->next = prev;
                prev = current;
                current = next;
            }

            ListNode *temp = groupPrev->next;
            groupPrev->next = kth;
            groupPrev = temp;
        }

        return dummy.next;
    }
};

File: 25. Reverse Nodes in k-Group/main.py
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverse(self, values: List, left: int, right: int):
        while left < right:
            values[left], values[right] = values[right], values[left]
            left += 1
            right -= 1

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        values = list()
        node = head

        while node:
            values.append(node.val)
            node = node.next

        size = len(values)

        for i in range(0, size - k + 1, k):
            self.reverse(values, i, i + k - 1)

        node = head
        for i in range(size):
            node.val = values[i]
            node = node.next

        return head


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or k <= 1:
            return head

        current = head

        for i in range(k):
            if not current:
                return head
            current = current.next

        current = head
        prev = None
        next = None
        count = 0

        while current and count < k:
            next = current.next
            current.next = prev
            prev = current
            current = next
            count += 1

        if next:
            head.next = self.reverseKGroup(next, k)

        return prev


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def getKth(self, node: Optional[ListNode], k: int) -> Optional[ListNode]:
        while node and k > 0:
            node = node.next
            k -= 1

        return node

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        groupPrev = dummy

        while True:
            kth = self.getKth(groupPrev, k)
            if not kth:
                break

            groupNext = kth.next
            prev, current = kth.next, groupPrev.next
            while current != groupNext:
                next = current.next
                current.next = prev
                prev = current
                current = next

            groupPrev.next, groupPrev = kth, groupPrev.next

        return dummy.next
